### PR TITLE
[3.3.5] Core/Movement: MotionMaster & MovementGenerators cleaning

### PR DESCRIPTION
**Pull request by ccrs**
[//]: # (***************************)
[//]: # (** FILL IN THIS TEMPLATE **)
[//]: # (***************************)

First stage of movement revamp: cleaning, renaming and refactor.

No real functionality change in this one.

### CODE REVIEW

**Comment by ccrs** [comment]
@Keader its by far not finished, but with this already you will notice a great improvement on those valks

**Comment by Aokromes** [comment]
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Movement/MovementGenerators/TargetedMovementGenerator.cpp:196:13: fatal error: 
      use of undeclared identifier 'i_path'
    return (i_path) ? (i_path->GetPathType() & PATHFIND_NORMAL) : true;
            ^
1 error generated.
make[2]: *** [src/server/game/CMakeFiles/game.dir/Movement/MovementGenerators/TargetedMovementGenerator.cpp.o] Error 1

**Comment by Aokromes** [comment]
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Movement/MovementGenerators/TargetedMovementGenerator.cpp:246:34: fatal error: 
      use of undeclared identifier 'GetTarget'
    return owner->GetVictim() != GetTarget();
                                 ^
1 error generated.
make[2]: *** [src/server/game/CMakeFiles/game.dir/Movement/MovementGenerators/TargetedMovementGenerator.cpp.o] Error 1

xd

**Comment by Keader** [comment]
She should stop blink after stuns? xd

**Comment by ccrs** [comment]
VS is trolling me really hard right now, I have no compile errors O.o

**Comment by tkrokli** [comment]
@ccrs : if you look at AppVeyor (vs2015), it does not detect undeclared identifiers either. Travis-ci (Linux, Clang) does.

**Comment by ccrs** [comment]
It's not undeclared, its inherited from ``TargetedMovementGeneratorBase``
Eitherway I'm updating right now to VS2017 while looking into solving it

**Comment by ccrs** [comment]
WaypointMovementGenerator will be left outside this since I'm planning to restore partially 28050f338dfc66e0c40b6a3915bf96e38e0613e5

**Comment by Shauren** [comment]
No, you should not try to bring back 28050f3, it was determined to be not blizzlike.

**Comment by ccrs** [comment]
More of a "this doesnt work... revert! D:"

But parts of it included some cleaning and refactoring would be nice to recover.

**Comment by Shauren** [comment]
No, it wasnt blizzlike AND didnt work, thats way more than enough to revert

**Comment by jackpoz** [comment]
If a refactor changes more than 200 lines/10 files it becomes pretty much unreviewable. None can ensure simply by looking at the code if there is any issue. In a way it takes less time to just do manually all the changes again.

**Comment by ccrs** [comment]
"In a way it takes less time to just do manually all the changes again." what do you mean?

**Comment by jackpoz** [comment]
The review of the refactor takes more time than doing the refactor again and committing it.

**Comment by ccrs** [comment]
Then I'll do some more test in couple hours and if everything goes ok I'll just commit

File: src/server/game/Movement/MovementGenerators/RandomMovementGenerator.cpp [hunk 1]
```diff
@@ -17,164 +17,133 @@
...
+    Position position(_reference);
+    float distance = frand(0.f, 1.f) * _wanderDistance;
+    float angle = frand(0.f, 1.f) * float(M_PI) * 2.f;
+    owner->MovePositionToFirstCollision(position, distance, angle);
 
-    //Call for creature group update
-    if (creature->GetFormation() && creature->GetFormation()->getLeader() == creature)
-        creature->GetFormation()->LeaderMoveTo(destX, destY, destZ);
-}
+    uint32 resetTimer = urand(1000, 2000);
```
Comment (jackpoz) [inline src/server/game/Movement/MovementGenerators/RandomMovementGenerator.cpp]: 50% of the time this urand() call is useless

File: src/server/game/Movement/MotionMaster.cpp [hunk 2]
```diff
@@ -88,9 +88,6 @@ void MotionMaster::UpdateMotion(uint32 diff)
     if (!_owner)
         return;
 
-    if (_owner->HasUnitState(UNIT_STATE_ROOT | UNIT_STATE_STUNNED)) // what about UNIT_STATE_DISTRACTED? Why is this not included?
-        return;
-
```
Comment (chaodhib) [inline src/server/game/Movement/MotionMaster.cpp]: it should be the other way around imo: if this condition is checked here, it should be removed from the MoveGens. This reduce code duplication.

Comment (ccrs) [inline src/server/game/Movement/MotionMaster.cpp]: Each movenment decides what to do when those states are detected.

If this is checked to prevent update on MotionMaster, that is no longer an option and its the reason why I removed it.

File: src/server/game/Movement/MovementGenerators/ConfusedMovementGenerator.cpp [hunk 3]
```diff
@@ -17,91 +17,94 @@
...
-bool ConfusedMovementGenerator<T>::DoUpdate(T* unit, uint32 diff)
+bool ConfusedMovementGenerator<T>::DoUpdate(T* owner, uint32 diff)
 {
-    if (unit->HasUnitState(UNIT_STATE_ROOT | UNIT_STATE_STUNNED | UNIT_STATE_DISTRACTED))
-        return true;
+    if (!owner || !owner->IsAlive())
+        return false;
 
-    if (i_nextMoveTime.Passed())
+    if (owner->HasUnitState(UNIT_STATE_NOT_MOVE) || owner->IsMovementPreventedByCasting())
```
Comment (chaodhib) [inline src/server/game/Movement/MovementGenerators/ConfusedMovementGenerator.cpp]: It's not really specific to your PR and it's more like a rant than anything else but: it would be great one day to remove this concept of "casting" from the MoveGens. a casting creature should stay still, not because it's casting a spell but because its AI is saying to its legs (MoveGens) "do not move". The legs (the MoveGens) shouldn't know the concept of "cast" or "spell".

In concrete terms, this means that when the AI starts to cast a spell, it should turn on a unit state (or a movementflag ?) like UNIT_STATE_NOT_MOVING_BY_AI that should be included in the UNIT_STATE_NOT_MOVE mask.

Comment (ccrs) [inline src/server/game/Movement/MovementGenerators/ConfusedMovementGenerator.cpp]: That can be done adding checks to ````IsMovementPreventedByCasting```` using a new state or any other option we can think of. The perfect move would be to change the name of the method into simply ````IsMovementPrevented````.

Comment (chaodhib) [inline src/server/game/Movement/MovementGenerators/ConfusedMovementGenerator.cpp]: Yes. That would be a good start in the right direction I think. At term though, it would be even better than the MoveGens do not call any methods of the unit or its AI. The AI should know the movement code and call it when it needs to move the unit its controlling. The movement code shouldn't be aware of the AI, nor of the unit as far as possible.

This decoupling will help us by making things simpler and easier to manage I think ([divide and conquer strategy](https://effectivesoftwaredesign.com/2011/06/06/divide-and-conquer-coping-with-complexity/)).

**Comment by chaodhib** [comment]
Well, that's a lot of changes (in term of removed/added LOCs). Would have been nice to have some explanation on what changes you did per MoveGen to able to review more quickly.

**Comment by ccrs** [comment]
What I did is pretty much refactor them so they all follow the same pattern and adapt it to the specifications of the generator itself.

The best way to compare IMO is to open two windows, for example:
https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Movement/MovementGenerators/ConfusedMovementGenerator.cpp
https://github.com/ccrs/TrinityCore/blob/movement/src/server/game/Movement/MovementGenerators/ConfusedMovementGenerator.cpp

Im confident this wont break anything related, if only, I fear something like a typo or similar.

File: src/server/game/Movement/MovementGenerators/PointMovementGenerator.cpp [hunk 4]
```diff
@@ -75,13 +75,13 @@ bool PointMovementGenerator<T>::DoUpdate(T* owner, uint32 /*diff*/)
...
     }
 
-    owner->AddUnitState(UNIT_STATE_ROAMING_MOVE);
-
     if ((_interrupt && owner->movespline->Finalized()) || (_recalculateSpeed && !owner->movespline->Finalized()))
     {
         _recalculateSpeed = false;
         _interrupt = false;
 
+        owner->AddUnitState(UNIT_STATE_ROAMING_MOVE);
```
Comment (chaodhib) [inline src/server/game/Movement/MovementGenerators/PointMovementGenerator.cpp]: changes not in right commit

Comment (ccrs) [inline src/server/game/Movement/MovementGenerators/PointMovementGenerator.cpp]: what do you mean?

Comment (tkrokli) [inline src/server/game/Movement/MovementGenerators/PointMovementGenerator.cpp]: @chaodhib : do you mean that this change belongs to a different PR, or that this line (and maybe lines below) belong in a different file (or part of this file) where a different commit was placed?

Comment (chaodhib) [inline src/server/game/Movement/MovementGenerators/PointMovementGenerator.cpp]: Thoses changes concerned the PointMovementGenerator but were in the Targeted MoveGen commit. Ccrs already fixed it though.

File: src/server/game/Movement/MotionMaster.cpp (hunk 2 above)
Comment (ariel-) [inline src/server/game/Movement/MotionMaster.cpp]: > Each movenment decides what to do when those states are detected.

Sounds like a good place to add a virtual function

Comment (ccrs) [inline src/server/game/Movement/MotionMaster.cpp]: Yeah it is, planning to do it
